### ISSUE TITLE
enums in go bindings should have the enum type

### ISSUE THREAD

**Issue by jynnantonix** [domokit/mojo#765]
Consider the following enum in a mojom:
```
enum Error {
  OK = 0,
  INVALID_ARGUMENT,
  PERMISSION_DENIED,
  BUSY,
  OUT_OF_BOUNDS,
  UNIMPLEMENTED,
  INTERNAL,
};
```

The generated go bindings are:

```go
type Error int32

const (
	Error_Ok = 0
	Error_InvalidArgument = Error_Ok + 1;
	Error_PermissionDenied = Error_InvalidArgument + 1;
	Error_Busy = Error_PermissionDenied + 1;
	Error_OutOfBounds = Error_Busy + 1;
	Error_Unimplemented = Error_OutOfBounds + 1;
	Error_Internal = Error_Unimplemented + 1;
)
```

While it's good that the bindings define the enum to be its own type, they don't actually define the values of the enum to have the same type.  This means that code like the following:

```go
func DoSomething() Error {
	err := Error_Ok

	// Do some work, maybe change err.

	return err
}
```

fails to compile with the error `cannot use err (type int) as type Error in return argument`.

This can be fixed by giving the enum values the enum type in the `const()`.  This would also make the 'Error_' prefix unnecessary since the compiler would check the actual types of the values.  The proper generated bindings should look like:

```go
type Error int32

const (
	Ok Error = 0
	InvalidArgument = Ok + 1;
	PermissionDenied = InvalidArgument + 1;
	Busy = PermissionDenied + 1;
	OutOfBounds = Busy + 1;
	Unimplemented = OutOfBounds + 1;
	Internal = Unimplemented + 1;
)
```

@vardhan-cr @crawshaw @shinykitten @MitchRudominer

**Comment by MitchRudominer** [comment]
Thanks this definitely seems wrong and I didn't know about it. Regarding the name prefix: this still seems useful to avoid polluting the package namespace.

**Comment by MitchRudominer** [comment]
Also the +1s can be avoided using iota

**Comment by jynnantonix** [comment]
A package is pretty narrowly scoped and any usage of the enum in another package requires that it be prefixed with the package name.  The only issue with removing the prefix is if two different enums in the same mojom have the same name but that can be solved by prefixing in the mojom itself.  So I don't really see the benefit to prefixing.

At the very least, we should at least get rid of the underscores.  Effective Go discourages using underscores and almost no real-world go packages use them.  golint also complains if it sees any underscores.  We should try to keep the generated code as idiomatic as possible.

**Comment by MitchRudominer** [comment]
I agree with getting rid of the underscores.  But the name collision between enums problem is real. I don't think it is sufficient to say that the user needs to do collision avoidance in the .mojom file because the Mojom language explicitly uses the enum type as a namespace for the enum values. So how about:
```
const (
    ErrorOk Error  = iota
    ErrorInvalidArgument
    ErrorPermissionDenied
    ErrorBusy
    ErrorOutOfBounds
    ErrorUnimplemented 
    ErrorInternal
)
```

**Comment by vardhan-cr** [comment]
the const typing you're proposing lgtm, but I have reservations about removing the enum name from the enum values in the generated go bindings:

I think we should prefix the enum values with the enum's name /because/ of the collision; enum value should always be enum-scoped, not namespace/module-scoped.  Retrofitting enum values to avoid collision with a different enum value (or any other identifier actually) in the same mojom scope (or module scope) is not a good solution. C++ and Dart bindings already deliberately scope their enum values with the enum name for this reason (i.e., in C++, you /have/ to do Error::Ok since it's an 'enum class').

Regards to the '_' in Go names; i think this is a general problem of trying to map mojom types (and features) to a language's types and features. AFAIK, its not possible to scope constants inside other types in Go, so I see this as a sacrifice. We can choose to name-mangle however we want (we currently managle mojom names of 'a_bucket' to Go names 'aBucket'), though I think we should keep the '_' in enum names.

**Comment by jynnantonix** [comment]
Right now every mojom file is its own go package.  I don't think it's too much to ask for mojom authors to avoid conflicts in the same file.  The choice is between unnecessary redundancy (`errors.Error_Ok` for the example from the first post) or conflicts in the same mojom.  I would rather avoid the unnecessary redundancy since that affects everyone that uses the enums instead of avoiding conflicts, which can only happen to a subset of the people who use enums.

@vardhan-cr My editor is full of linter errors everywhere I use something from the generated bindings that has underscores.  While I understand that it's hard to map mojom types and features to go features, I would strongly recommend against anything that conflicts so heavily with the established conventions for the language.

**Comment by krisgiesing** [comment]
My only concern with asking mojom authors to do the conflict avoidance is language-specific gotchas. That is, if the conflict avoidance is only necessary in go, then mojom authors who primarily use other languages may inadvertently paint themselves into a corner when they start to use go, and have to back up and fix a bunch of stuff.

**Comment by MitchRudominer** [comment]
The fact of the matter is that
```
enum A {
  X,
  Y
};

enum B {
  X,
  Y
};
```
Is valid according to the Mojom spec. So it is not a matter of best practices. The job of the Mojom compiler is to translate valid mojom into valid C++, Go, Dart, etc.  So unless we change the Mojom spec to disallow the above construct (which I don't think we should do) then some name mangling scheme in Go will be necessary.

Another thing to keep in mind is that it is an implementation detail that there is a one-to-one correspondence between a .mojom file and a Go package. You should not think of a .mojom file as necessarily coresponding to a Go package.

**Comment by jynnantonix** [comment]
Well my preference would be to change the mojom spec ;-) but I realize that's not a popular opinion.  So prefixing (without underscores) sgtm.

**Comment by vardhan-cr** [comment]
BTW, I don't have any strong opinions on this keeping the _ (just a preference); just making my case..

Regarding your lint error (and assuming you are scoping your enum values); that's unfortunate, i aknowledge that's a valid concern.  Sure, you could avoid the _ to be more language idiomatic, but once you have an enum inside a struct (which is valid), you will have enum values styled as 'MyStructMyEnumMyValue' or something, which looks ugly to me;  where does the struct's name end and enum's name begin?
